ui/raidboss/emulator/data/network_log_converter/ParseLine.js

```diff
@@ -11,13 +37,83 @@ class ParseLine {
     if (event === '252')
       return false;
 
-    // This is ugly, but just want to check if we have a specific
-    // line class is defined without having a huge if/else or switch/case
-    let tn = 'LineEvent' + event;
-    if (eval('typeof(' + tn + ')') === 'function')
-      eval('ret = new ' + tn + '(repo, line, parts);');
-    else
+    // This is ugly, but Webpack prefers being explicit
+    switch ('LineEvent' + event) {
```

Comment:
How about using [dynamic imports](https://webpack.js.org/api/module-methods/) for this?

Reply:
Okay, I finally got around to trying this. Keep in mind I'm coming at this with zero expertise, so please bear with me if I did something very not-what-you-expected. I removed all the imports and the switch, and put back the code with the following modifications (no idea why the hex conversion was giving me so much trouble, but that ended up working with the configuration below):
```diff
    // This is ugly, but just want to check if we have a specific
    // line class is defined without having a huge if/else or switch/case
    let tn = 'LineEvent' + event;
+   const hexEvent = parseInt(event, 10).toString(16);
+   const module = import(`./LineEvent0x${hexEvent}.js`);
    if (eval('typeof(' + tn + ')') === 'function') {
      eval('ret = new ' + tn + '(repo, line, parts);');
    } else {
      ret = new LineEvent(repo, line, parts);
    }
```

My problem here now is very similar to my problem with other `eval()`s in this patch: they don't know what they're looking for. I can `import X from 'Y.js';` and the downstream `eval()` won't know what X is. My workaround has been to `import X as _X from 'Y.js'; const X = _X;` If there's a better way to approach this, I'd love to know; I personally feel that the workaround is verging on disgusting.

As for this, I don't know of a way to ``const <insert dynamic name here> = import(`./LineEvent0x${hexEvent}.js`);``. My initial thought was to shove it into a map via `[varName]` syntax, but that doesn't stop the fact that at some point I need to retrieve it explicitly named the name that the `eval()` wants it to be:
```js
    // perfect code above this line
    if (importMap.has(tn)) {
      const LineEvent04 = importMap.get(tn); // How can I know this? How can I get LineEvent05, LineEvent39, etc.?
      eval('ret = new ' + tn + '(repo, line, parts);');
    }
    // perfect code below this line
```

(Some attempts later, I got the above to work via object destructuring and renaming, but the class doesn't know about the class that it extends. D'oh.)

Additionally, I found that this was creating [chunks](https://webpack.js.org/guides/code-splitting/), which seems kind of painful for distributing to people; I think sending 24 additional bundles is kind of a non-starter:
```
0.bundle.js  8.03 KiB             0  [emitted]
1.bundle.js  5.17 KiB             1  [emitted]
. . .
23.bundle.js  4.39 KiB             2  [emitted]
raidemulator.bundle.js   780 KiB  raidemulator  [emitted]  raidemulator
```

I was able to get around this by adding:
```js
  plugins: [
    new webpack.optimize.LimitChunkCountPlugin({
      maxChunks: 1,
    }),
  ],
```
to the Webpack config, but this all seems like a lot to be more clever with the code.

I could be wrong about all of the above; I tried to give it what I felt like was a fair attempt, but I can't find working implementation, and the paths that I've been going down have had more dragons than UCoB. Please correct me if I'm doing something very dumb here!

Reply:
`import()` returns a promise, you could actually do something like
```js
    const hexEvent = parseInt(event, 10)
      .toString(16)
      .toUpperCase()
      .padStart(2, "0");
    let LineEvent;
    try {
      const module = await import(
        // eager so it doesn't generate an extra chunk
        /* webpackMode: "eager" */ `./LineEvent0x${hexEvent}.js`
      );
      if (!Object.prototype.hasOwnProperty.call(module, hexEvent)) {
        throw void 0;
      }
      LineEvent = module[hexEvent];
    } catch (_) {
      ({ default: LineEvent } = await import(
        /* webpackMode: "eager" */ `./LineEvent.js`
      ));
    }
```
Then do a `Promise.all` in the `NetworkLogConverter`
```js
lines = (await Promise.all(lines.map((l) => ParseLine.parse(repo, l)))).filter((l) => l);
```
But I think a much cleaner approach would be to make a dict/map of log lines and just import it?
maybe 
```console
└───network_log_converter
    │   LogRepository.js
    │   ParseLine.js
    │
    └───events
            0x00.js
            ...
            index.js
```
`import` all the events inside index and return an object/dict and simply access them using a property.

Reply:
Another option to abstract it a bit since this is a class anyway.. :unamused: 
```js
export default class LineEventFactory {
  constructor(event, props) {
    switch (event) {
    case "00":
      return new LineEvent00(...props);
    case "01":
      return new LineEvent01(...props);
    // ...
    default:
      return new LineEvent(...props);
    }
  }
}
```
```js
ret = new LineEventFactory(event, [repo, line, parts]);
```

Reply:
Okay, so, in fairness, I haven't tried these. I think all of these (I'd have to see about the first one, seems a little magic to me) are better than the current/proposed implementation.

I think I want to hold off on making changes that actually change the code until after this lands, at which point either you or I (it's your code!) can make any one of these follow-up changes. In all reality, I should have left the eval() since I leave it everywhere else and just put that horrid `TODO` banner, but this felt small enough to just nip it and then move on. I think the switch statement is ugly, but there's no cleverness, so I think most people can agree from a passing glance that "this does what it was doing before" even if it's 100 more lines of code or whatever (🤮).

How do you feel? I guess I feel like I could be persuaded that "if I'm fixing the thing, might as well fix it good" but I'm also currently of the opinion that "yea, 100+ files is a fine enough PR for now"